### ISSUE TITLE
Cleanup in messages.proto and other public APIs

### ISSUE THREAD

**Issue by sashafrey** [bigartm/bigartm#356]
For prototype see PR #426 .

This page describes upcoming changes in [messages.proto](https://github.com/bigartm/bigartm/blob/master/src/artm/messages.proto) and other  public API of BigARTM library. Most changes are cleaning legacy stuff or doing consistency fixes to make APIs clean and shiny. Overall, the impact of the changes on `ARTM.py` (python interface) and `BigARTM.exe` (command-line interface) is minimal, because those APIs already use only 'good' features that are not affected by this cleanup. However, this changes might **not** be backwards compatible. Everyone who is using BigARTM should familiarize himself with this changes, and is highly welcome to ask questions or leave an opinion.

Before we go into a really long list of what is going to change let's list few bad decisions that will **NOT** change. The reason is that we either had no time to design the appropriate change, or it might require too much renaming. As a result of renaming the API and the internal code of the library might became inconsistent and hard to follow.
* These change is not concerned about python API (`ARTM.py`) and `BigARTM.exe` (CLI). The goal is, in fact, to keep changes in `ARTM.py` and `BigARTM.exe` at minimum, because those APIs were designed recently and are quite good. Instead, we in this issue are focusing on changes to `messages.proto`, `c_interface.h` (low-level API) and `cpp_interface.h` (C++ API). 
* We plan to keep term `class_id` whenever we specify modality (aka namespace) of a token.
* We plan to keep the term `name` in `model_name`, `topic_name`, etc..., even though this refers to an identifier so it would be more natural to call this `id`. In software there are other cases when `name` refers to an identifier. For example, in Windows Forms designer in Visual Studio the `name` of a component is an identifier, which can be used in the source code to refer to the component.
* For now we will keep `MasterComponent.cache_theta` and `ProcessBatchesArgs.reuse_theta` flags. They are designed for offline algorithm. An alternative way to implement offline algorithm is represent theta matrix via an adding a new modality (aka `class_id`).

We still need to **decide** on the following questions:
* What naming should we define for disk locations? (`path`, `file`, `filename`,`folder`, etc...)
* How should we fix inconsistencies between `num_xxx` and `xxx_count` fields?
* Should we ever use `optional DoubleArray` (same for other `optional XxxxArray` fields)?
* TBD: compact range of indices for all protobuf messages

Upcoming changes in messages.proto
============================

* [https://github.com/sashafrey/bigartm/commit/95e78a83f3ad2a15bae5c22a146b6c29cf2719a5] ~~Introduce `Item.token_id` and `Item.token_weight` to represent bag-of-words information.~~
* [Done by #462] [https://github.com/sashafrey/bigartm/commit/1a4d2a9a0e019185e1190c4b8772d6b8b459de88] ~~Remove message `Stream` and `MasterComponentConfig.stream`. No replacement will be provided for this feature. This feature is very weak and do not provide users with a way to handle train and test data. This needs to be done manually outside of the library. The simplest way would be to move some batches into a separate folder. Users may also prepare two text files in VW or UCI-BOW formats and parse them separately.~~
* [https://github.com/sashafrey/bigartm/commit/aa4f5536cfc95a9e79fbc92c50b55c68b15faafc, https://github.com/sashafrey/bigartm/commit/2bc3af2218c053b3ec79caa01c45f7551a6ca8d5, https://github.com/sashafrey/bigartm/commit/33713d5a641b160f6c8f7bf0626cee05eeeca034] ~~Remove miscelaneous obsolete fields: `MasterComponentConfig.online_batch_processing`, `ModelConfig.score_name`, `PerplexityScoreConfig.theta_sparsity_*`, `ThetaSnippetConfig.item_id`, `TopicModel.internals`, `MasterComponentInfo.master_id`, `TopicModel.operation_type`. `MasterComponentConfig.disk_path`, `MasterComponentConfig.compact_batches`, `TopicModel.seed` (a hack that was introduced to simplify implementation);~~
* [Done by #462] ~~Remove `*.use_sparse_bow` (was an experimental feature that enabled dense operations in `ProcessBatch`; and it didn't fly - time to remove it). I believe nobody has ever used these fields. Also, remove message `BoolArray`.~~
* [https://github.com/sashafrey/bigartm/commit/a9768a1891bc760d36c37be7db0c8bfa6a87b78a] ~~Remove obsolete fields `CollectionParserConfig.cooc*` fields. Similar functionality is available through `GatherDictinoaryArgs` (a newer API).~~
* [https://github.com/sashafrey/bigartm/commit/1ba2a012f9e39ff4b35ca340e7dc916ca9649947] ~~Remove obsolete fields in `InitializeModelArgs`. Similar functionality is available through `FilterDictionaryArgs`. For `SourceType.Batches` one has to first use `GatherDictionaryArgs` to generate a dictionary, and then `InitializeModelArgs` with that dictionary.~~
* [https://github.com/sashafrey/bigartm/commit/11d8966582a1561da1d8e5bff3682a6babf6b944] ~~Remove `InitializeModelArgs.topics_count` field. For now users of the low-level API will have generate topic names in repeated `InitializeModelArgs.topic_name` field. This will be hidden at high-level APIs. Later we will provide `TopicSet` functionality ( #324 ). Remove field `GetThetaMatrixArgs.topic_index` (legacy stuff that existed before we even introduced the concept of `topic_name`.~~
* [Done in #455] [https://github.com/sashafrey/bigartm/commit/aecaf98066b6cf52cc056b179c328f7d05d165b6] ~~Remove fields `GetThetaMatrixArgs.batch` and `GetScoreValues.batch`. Instead the users are expected to run ProcessBatches operation that can return the same information.~~
* [Done in #456, #455] [https://github.com/sashafrey/bigartm/commit/f2022d9aa1ba32948f7f74e62752416767a57e4e] ~~`ProcessBatchesArgs.reset_scores` and `GetThetaMatrixArgs.clean_cache` should be replaced with a separate API function that performs the action.~~
* [https://github.com/sashafrey/bigartm/commit/edbafe6aff9121510f64d85cbebe20d4b93decd0] ~~Remove obsolete ImportBatchesArgs.batch_name. `Batch.id` of the imported batches will be used as an identifier.~~
* [https://github.com/sashafrey/bigartm/commit/1bd4150d2ceed0f9058d1a73f730f88718ee36ad] Change to improve consistency around `RequestType`. See the commit description for more info.
* ~~`MasterModelConfig.threads` should be renamed to `MasterModelConfig.processors_count` for consistncy with `MasterComponentConfig`. Even though this is named `threads` in BigARTM CLI I believe `processors_count` is a better name for a protobuf message.~~

Upcoming changes in Scores and Regularizers
=================================
* [https://github.com/sashafrey/bigartm/commit/d6eec5d07aacc0c09487afebf59cda6a602e6242] `RegularizerConfig` as a field in `MasterComponentConfig`. We remove all APIs to operate on individual regularizers (`ArtmCreateRegularizer`, `ArtmReconfigureRegularizer`, `ArtmDisposeRegularizer`). Instead users are supposed to change `regularizer_config` field in `MasterComponentConfig`, and use "ArtmReconfigureMasterComponent" to apply these changes. This is how it works for `MasterModel`, and this is also the way it works for scores in `MasterComponent`. Why regularizers in `MasterComponent` should be any different?
* [https://github.com/sashafrey/bigartm/commit/0ee00c5b228f2a295f056e8c356f2948ea20d1e5] Remove `RegularizerSettings.use_relative_regularization`. To enable relative regularization it should be enough to set `RegularizerSettings.gamma` parameter. Otherwise (when `gamma` is not set) the old ('absolute') regularization coefficients will be used. Move fields `tau`, and `gamma` from `RegularizerSettings` into `RegularizerConfig` message. Remove `RegularizerSettings` message. Remove `ProcessBatchesArgs.regularizer_tau`. This imply that all models within one master component will share the same regularization coefficient. Changing regularization coefficient needs to go through regularizer reconfiguration.
* [https://github.com/sashafrey/bigartm/commit/ff3fde0d916d4bc3db5c8008213bc80a7293a6ea] Move `Type` enums outside of the messages. Merge enums `ScoreData.Type` and `ScoreConfig.Type` into `ScoreType` enum. Move `RegularizerConfig.Type` into `RegularizerType enum`. Rename fields `ScoreData.type`, `ScoreConfig.type` and `RegularizerConfig.type` into `xxxx_type`. This is done to avoid build issues when generating protobuf messages in `C#`, where `type` is a keyword that can't be used in proto. I'm not 100% sure, but anyway it sounds like a good idea to avoid word `type` because it is too generic.
* [https://github.com/sashafrey/bigartm/commit/f1dd6d0db6af600efd04a52066b751425acaf1e2] Add `repeated string` fields `topic_name`, `dictionary_name`, `class_id` into `RegularizerConfig`. Remove corresponding fields from all specific score configs and regularizer configs.
* [https://github.com/sashafrey/bigartm/commit/2cf63d4bde64f02699aafdcd530ba391dd79cd1b] Add `optional float eps [default = 1e-37]` field into `ScoreConfig` message. This is a threshold to be used when numbers are compared with zero. This is common to several specific `XxxxScoreConfig` messages and should be promoted to `ScoreConfig`. `XxxxScoreConfig.eps` field should be removed.
* [https://github.com/sashafrey/bigartm/commit/3d3b474989cd04f1c7a0017e9805bc5ad137c875] Add `ScoreData.value`, `ScoreData.numerator` and `ScoreData.denominator`. This is because the `value` is calculated as a fraction `a/b`, and many scores want to expose this information to the user. By having a common place to store `value`, `numerator` and `denominator` we value remove the following: `PerplexityScore.value, raw, normalizer`, `SparsityThetaScore.value, zero_topics, total_tokens`, `SparsityPhiScore.value, zero_topics, total_tokens`, `TopicMassPhiScore.value`, `ClassPredictionScore.value, error, total`.
* [https://github.com/sashafrey/bigartm/commit/cee90d2d1e323dee48e4f38fc1baaae9372b9de9] Remove `PerplexityScoreConfig.type`. `UnigramCollectionModel` should be enabled whenever user provides a dictionary, otherwise the only choice is to fall back to  `UnigramDocumentModel`.
* [https://github.com/sashafrey/bigartm/commit/f61fa3fa143793d95cc0477372c849eacf58e2f5] Remove empty messages (`XxxxScoreConfig`, `XxxxScore`, `XxxxRegularizerConfig`). Also remove `ItemsProcessed` score because it is kind of useless...
* [https://github.com/sashafrey/bigartm/commit/aa0e001715b6fc9bc3d71dd49069e5c9b316944b] Add `TopicKernelScoreConfig.type`, which can be `Size`, `Purity`, `Contrast` and `Coherence`. To retrieve multiple users need to created several scores. The `average_xxxx` values will be written into `ScoreData.value`.
* [Done in #455][https://github.com/sashafrey/bigartm/commit/3fa206431dfa08e5800764af66dd886d574d0739] ~~Remove API to retrieve regularizer state~~.

Upcoming changes in C++ API 
======================
* [Done in https://github.com/bigartm/bigartm/pull/451] [https://github.com/sashafrey/bigartm/commit/77b400cf79d9d7234f5b4b6d368341249dbfebda] ~~Everything except `MasterModel` gets removed. For now all low-level building blocks (ProcessBatches, Merge, Regularize, Normalize) will NOT be exposed in C++ API. Users are expected to work with `MasterModel` API, introduced recently (https://github.com/bigartm/bigartm/issues/325).~~

Remove old style models
===================
* [Done in #455] [https://github.com/sashafrey/bigartm/commit/b937b55fbe7f204d4f14bc9d4f1ec2452ea0e56b] ~~commit removes support for old-style models that worked via `ArtmInvokeIteration` (or `ArtmAddBatches`) / `ArtmWaitIdle` / `ArtmSynchronize` APIs. This commit also removes the corresponding `XxxArgs` from protobuf messages, and removes `model-v06` option from BigARTM CLI.~~
* [Done in #462] ~~[https://github.com/sashafrey/bigartm/commit/6e8e7c28d6d0427641243df134de60851d528564] Remove `ModelConfig` message (except removal of `ArtmOverwriteTopicModel`, which will be kept).~~

Upcoming changes in `c_interface.h`
===============================
Look [here](https://github.com/bigartm/bigartm/pull/426/files#diff-f2edda319ba7de36830b225165259dc7) to see all planned changes in `c_interface.h`. Most changes are described above, here we just briefly re-iterate:
* `ArtmCreateModel`, `ArtmReconfigureModel`, `ArtmAddBatch`, `ArtmInvokeIteration`, `ArtmWaitIdle`, `ArtmSynchronizeModel` are removed because they deal with old-style models.
* `ArtmCreateRegularizer`, `ArtmReconfigureRegularizer`, `ArtmDisposeRegularizer` are removed because `RegularizerConfig` became part of `MasterComponentConfig`. Users should use `ArtmReconfigureMasterComponent` to apply changes to regularizers.
* `ArtmOverwriteTopicModel` is confusing and partially overlaps with `ArtmAttachModel` functionality. The later API should be used to change value in `pwt`, `nwt` or `rwt` matrices. Changing the structure of the model (e.g. adding/removing tokens or topics) is still an open issue which can partially solved via `ArtmMergeModel`, and partially by creating a new matrix.
* [Done in #455] ~~`ArtmRequestRegularizerState` is removed, and no alternative API is provided. The issue is that this API had been in place for more than a year, and noone ever used it. Apparently this feature can wait until someone makes a request.~~
* [https://github.com/sashafrey/bigartm/commit/52a280f37d2dfcf4e5be8179524597e3fd735643] Remove `ArtmSaveBatch` and `ArtmRequestLoadBatch`, because the user may simply dump a serialized protobuf message to disk. For example,
```
# in Python
batch = artm.messages.Batch()
with open(batch_name, "rb") as f:
    batch.ParseFromString(f.read())

# in C++
Batch batch;
std::ifstream fin(batch_name, std::ifstream::binary);
batch.ParseFromIstream(&fin);
```

Stuff that we decided to keep
========================
The following changes will NOT be made:
* [https://github.com/sashafrey/bigartm/commit/60d1acfeeed74e0b442cbf3a2ea7bb20b3af219c] Remove obsolete field *.opt_for_avx. This actually removes a partially useful feature. Read more in the description of the commit.
* [https://github.com/sashafrey/bigartm/commit/6e8e7c28d6d0427641243df134de60851d528564] 
The part about removing `ModelConfig` message is still valid, however we will NOT remove `ArtmOverwriteTopicModel`. This API is still useful, even though `ArtmAttachModel` provides an alternative API to edit values in `pwt`, `nwt` or `rwt` matrices, and `ArtmMergeModel` provides an API to change the set of tokens and in the future it might also handle changes of `topic_name` (topical structure of the model).
* Remove field `TopicModel.topics_count`, `ThetaMatrix.topics_count`. Users should use the size of a repeated field `topic_name`. 
* In the`Batch` format we kept `Field` for backward compatibility. It is obsolete and users shell not use it; isntead, use `Item.token_id` and `Item.token_weight`.

**Comment by sashafrey** [comment]
@romovpa Please review this plan :) Any comments on the the questions the still need a decision?

**Comment by AnyaP** [comment]
Some thoughts on the questions that still need decision:
1. If you represent Theta as a new modality (class_id) will it be convenient to normalize it everywhere using the other dimension (we need p(t|d) but not p(d|t) as for any other modality). Or you mean some new mechanism just similar to the modalities but not equivalent?
2. I think that the final user in IPython API should be able to ask for 100 topics without creating their names. Or for 90 domain + 10 background topics. However, I have no thoughts on where it's better to translate the wishes of the user into generated names, if topics_count field is deleted.

**Comment by sashafrey** [comment]
@AnyaP
1. For this change we will not represent `Theta` as an extra modality. We can discuss this feature later. 2. Agreed :) `topics_count` will be removed from low-level API, but it will stay in high-level APIs. Later we will add the ability to use topic sets ( #324 ) in low-level APIs.

**Comment by sashafrey** [comment]
The remaining work from this ticket had been moved to https://github.com/bigartm/bigartm/pull/402 and https://github.com/bigartm/bigartm/issues/547
